blockchain/state_migration.go

```diff
@@ -371,3 +373,134 @@ func (bc *BlockChain) StopStateMigration() error {
 }
+
+func (bc *BlockChain) concurrentIterateTrie(root common.Hash, db state.Database, result chan common.Hash, finish chan error) (resultErr error) {
+	defer func() {
+		finish <- resultErr
+	}()
+
+	stateDB, err := state.New(root, db)
+	if err != nil {
+		return err
+	}
+
+	it := state.NewNodeIterator(stateDB)
+	for it.Next() {
+		result <- it.Hash
+
+		select {
+		case <-bc.quitWarmUp:
+			return errors.New("quitWarmUp")
+		case <-bc.quit:
+			return errors.New("quit")
+		default:
+		}
+	}
+
+	return nil
+}
+
+func (bc *BlockChain) StartWarmUp() error {
+	if bc.quitWarmUp != nil {
+		return fmt.Errorf("already warming up")
+	}
+
+	block := bc.GetBlockByNumber(bc.lastCommittedBlock)
+	if block == nil {
+		return fmt.Errorf("block #%d not found", bc.lastCommittedBlock)
+	}
+
+	mainTrieDB := bc.StateCache().TrieDB()
+	mainTrieCacheLimit := uint64(mainTrieDB.GetTrieNodeCacheLimit())
+	db := state.NewDatabaseWithCache(bc.db, mainTrieDB.TrieNodeCache())
+
+	bc.quitWarmUp = make(chan struct{})
+
+	go func() {
+		defer func() {
+			bc.quitWarmUp = nil
+		}()
+
+		var stats fastcache.Stats
+		var context []interface{}
+		var percent uint64
+		var cnt int
+
+		updateContext := func() []interface{} {
+			stats = fastcache.Stats{}
+			mainTrieDB.TrieNodeCache().UpdateStats(&stats)
+			percent = stats.BytesSize * 100 / mainTrieCacheLimit
+			context = []interface{}{
+				"warmUpCnt", cnt,
+				"cacheLimit", units.Base2Bytes(mainTrieCacheLimit).String(),
+				"cachedSize", units.Base2Bytes(stats.BytesSize).String(),
+				"percent", percent,
+			}
+			return context
+		}
+
+		children, err := db.TrieDB().NodeChildren(block.Root())
+		if err != nil {
+			logger.Error("Warm up is stop by err", "err", err)
+		}
+
+		logger.Info("Warm up is started", "blockNum", block.NumberU64(), "root", block.Root().String(), "len(children)", len(children))
+
+		var resultErr error
+		resultHashCh := make(chan common.Hash, 10000)
+		resultErrCh := make(chan error)
+		finishCnt := 0
+
+		for _, child := range children {
+			go bc.concurrentIterateTrie(child, db, resultHashCh, resultErrCh)
+		}
+
+		logged := time.Now()
+	loop:
+		for {
+			select {
+			case <-resultHashCh:
+				cnt++
+				if time.Since(logged) > log.StatsReportLimit {
+					logged = time.Now()
+
+					updateContext()
+					if percent > 90 { //more than 90%
+						close(bc.quitWarmUp)
+						logger.Info("Warm up is completed", context...)
+						return
+					}
+
+					logger.Info("Warm up progress", context...)
+				}
+			case err := <-resultErrCh:
+				if err != nil {
+					resultErr = err
+					logger.Warn("Warm up got an error", "err", err)
+				}
+
+				finishCnt++
+				logger.Debug("Warm up is being finished", "finishCnt", finishCnt, "err", err)
+
+				if finishCnt >= len(children) {
+					break loop
+				}
+			}
+		}
+		updateContext()
+		context = append(context, "resultErr", resultErr)
+		logger.Info("Warm up is completed", context...)
```

Comment:
```suggestion
		logger.Info("Warm up is completed", updateContext()..., "resultErr", resultErr)
```

Reply:
I tried but it doesn't work. ^^;

Reply:
How about this?
```
logger.Info("Warm up is completed", "resultErr", resultErr, updateContext()...)
```

Reply:
This also doesn't work.T.T
![image](https://user-images.githubusercontent.com/29390878/85240000-69598500-b471-11ea-9c08-537cae72a035.png)